File: aoc/solutions/_2019/day01.py
from aoc.utils.interfaces import Solution


def main(puzzle_input: list[str]) -> Solution:
    part1 = sum([fuel_mass(mass) for mass in map(int, puzzle_input)])
    part2 = sum([recursive_fuel_mass(mass) for mass in map(int, puzzle_input)])
    return Solution(part1, part2)


def fuel_mass(mass: int) -> int:
    return max(mass // 3 - 2, 0)


def recursive_fuel_mass(mass: int) -> int:
    fuel = fuel_mass(mass)
    if fuel <= 0:
        return 0
    return fuel + recursive_fuel_mass(fuel)

File: 2019/src/solutions/day01.rs
use crate::solutions::Solution;

pub fn run(puzzle_input: Vec<String>) -> Solution {
    let numbers: Vec<i32> = puzzle_input
        .into_iter()
        .map(|n| n.parse().unwrap())
        .collect();

    let part1: i32 = numbers.iter().map(|n| fuel_required(*n)).sum();
    let part2: i32 = numbers.into_iter().map(|n| recursive_fuel(n, 0)).sum();

    Solution { part1, part2 }
}

fn fuel_required(mass: i32) -> i32 {
    mass / 3 - 2
}

fn recursive_fuel(mass: i32, acc: i32) -> i32 {
    let fuel = fuel_required(mass);
    if fuel <= 0 {
        return acc;
    }
    recursive_fuel(fuel, acc + fuel)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fuel_part1() {
        assert_eq!(2, fuel_required(12));
        assert_eq!(2, fuel_required(14));
        assert_eq!(654, fuel_required(1969));
        assert_eq!(33583, fuel_required(100756));
    }

    #[test]
    fn fuel_part2() {
        assert_eq!(2, recursive_fuel(14, 0));
        assert_eq!(966, recursive_fuel(1969, 0));
        assert_eq!(50346, recursive_fuel(100756, 0));
    }
}
